python/cuml/linear_model/linear_regression_mg.pyx

```diff
@@ -23,610 +22,474 @@ import ctypes
+
+            arr_interfaces_y.append({"obj": y_m,
+                                     "data": input_ptr,
+                                     "shape": (n_rows, n_cols)})
+
+        n_total_parts = len(input_data)
+        cdef RankSizePair **rankSizePair = <RankSizePair**> \
+            malloc(sizeof(RankSizePair**)
+                   * n_total_parts)
+
+        for i in range(len(input_data)):
+            rankSizePair[i] = <RankSizePair*> \
+                malloc(sizeof(RankSizePair))
+            rankSizePair[i].rank = <int>rnk
+            rankSizePair[i].size = <size_t>len(input_data[i][0])
+
+        self.coef_ = cudf.Series(zeros(self.n_cols,
+                                       dtype=self.dtype))
+        cdef uintptr_t coef_ptr = get_cudf_column_ptr(self.coef_)
+
+        cdef float c_intercept1
+        cdef double c_intercept2
+        cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
+        cdef uintptr_t data
+        cdef uintptr_t labels
+
+        if self.dtype == np.float32:
+            data = self._build_dataFloat(arr_interfaces)
+            labels = self._build_dataFloat(arr_interfaces_y)
+
+            fit(handle_[0],
+                <RankSizePair**>rankSizePair,
+                <size_t> n_total_parts,
+                <floatData_t**>data,
+                <size_t>M,
+                <size_t>N,
+                <floatData_t**>labels,
+                <float*>coef_ptr,
+                <float*>&c_intercept1,
+                <bool>self.fit_intercept,
+                <bool>self.normalize,
+                <int>self.algo,
+                False)
+
+            self.intercept_ = c_intercept1
         else:
-            olsPredictSPMG(<double*>X_ptr,
-                           <int> n_rows,
-                           <int> n_cols,
-                           <double*>coef_ptr,
-                           <double>self.intercept_,
-                           <double*>pred_ptr,
-                           <int*>gpu_ids_ptr,
-                           <int>n_gpus)
-
-        return pred
-
-    def _fit_mg(self, alloc_info, params):
-
-        self.fit_intercept = params['fit_intercept']
-        self.normalize = params['normalize']
-
-        if alloc_info[0][0]['typestr'] == '<f8':
-            dtype = 'double'
+            data = self._build_dataDouble(arr_interfaces)
+            labels = self._build_dataDouble(arr_interfaces_y)
+
+            fit(handle_[0],
+                <RankSizePair**>rankSizePair,
+                <size_t> n_total_parts,
+                <doubleData_t**>data,
+                <size_t>M,
+                <size_t>N,
+                <doubleData_t**>labels,
+                <double*>coef_ptr,
+                <double*>&c_intercept2,
+                <bool>self.fit_intercept,
+                <bool>self.normalize,
+                <int>self.algo,
+                False)
+
+            self.intercept_ = c_intercept2
+
+        self.handle.sync()
```

Comment:
this is may be a misunderstanding on my part, but why don't you need to sync before re-reading c_intercept2 and storing it in another variable?

Reply:
c_intercept1 and c_intercept2 are host variables. So no need to sync before these two. Also, all the streams are synced before leaving the C++ anyways.